README.md
# holiday

A php class for determining holidays for many countries, regions and languages.
All defintions are in a small SQLite database that can be changed and expanded by the user.
The database "JspitHoliday.sqlite" currently supports the following countries

  "DE":Germany, "AT":Austria, "NL":Netherlands, "DK":Denmark, "FR":France,
  "IT":Italy, "ES":Spain(\*), "LU":Luxembourg, "BE":Belgium,
  "GR":Greece, "SK":Slovakia, "IE":Ireland, "CY":Cyprus, "PT":Portugal,
  "EE":Estonia, "FI":Finland, "LV":Latvia, "LT":Lithuania, "MT":Malta
  "CZ":Czech Republic, "PL":Poland, "CH":Switzerland, "GB":Great Britain(\*),
  "SE":Sweden, "BG":Bulgaria, "HR":Croatia, "RO":Romania, "SI":Slovenia,
  "HU":Hungary, "US":United States(\*),"JP":Japan,"RU:Russia

  (*) without regional holidays

 Database can be used without any claim to correctness and completeness.

## Usage

Include class JspitHoliday.php (1 File) directly with require or use a autoloader.
Put JspitHoliday.sqlite in the same Directory how JspitHoliday.php.
If you want to put the configuration JspitHoliday.sqlite in another directory, 
then specify the full path in the 2.Parameter of the constructor.

```php
<?php

$holiday = new JspitHoliday("DE-BB"); 
$holidayList = $holiday->holidayList(2018,'en');

```
$holidayList contain a array with all public holidays from 
country Germany(de) Region Brandenburg (bb) with english names(en) for the year 2018.

```php
array (
  '2018-01-01' => "New Year's Day",
  '2018-03-30' => "Good Friday",
  '2018-04-01' => "Easter",
  '2018-04-02' => "Easter Monday",
  '2018-05-01' => "Labor Day",
  '2018-05-10' => "Ascension of Christ",
  '2018-05-20' => "Whit Sunday",
  '2018-05-21' => "Whit Monday",
  '2018-10-03' => "Day of German Unity",
  '2018-10-31' => "Reformation Day",
  '2018-12-25' => "Christmas Day",
  '2018-12-26' => "Boxing Day",
)
```
Regions can be nested to depth 3. The notation is based on the ISO standard 3361.
Examples: "DE", "NL", "DE-BY", "DE-BY-SCH-A"
Last for Germany-Bavaria-Schwabing-Augsburg(City). 
The languages can be divided into dialects, e.g. 'de-DE' , 'de-CH'.

Get all holiday dates between $dateStart and $dateEnd which are from a Monday to Friday.

```php
$holiday = JspitHoliday::create('de-BE');
$dateArray = $holiday->dateList("2019-03-01","2019-03-31",array(1,2,3,4,5));
```

Further examples:

```php
$dateTime = new DateTime("1 May 2018 08:00");

$holidaysDE = JspitHoliday::create("DE");
if($holidaysDE->isHoliday($dateTime)) {
  echo "1 May 2018 is in DE a holiday";
}

//holidayName
$holidayName = $holidaysDE->holidayName('3 Oct','en');
if($holidayName) {
  echo $holidayName;
  //'Day of German Unity'
}

//holidayNameList
$holidaysIL = JspitHoliday::create("IL","JspitHoliday.sqlite");
$list = $holidaysIL->holidayNameList("Pessach I",2018,2022,'de');

var_dump($list);
/*
array(5) {
  ["2018-03-31"]=>
  string(9) "Pessach I"
  ["2019-04-20"]=>
  string(9) "Pessach I"
  ["2020-04-09"]=>
  string(9) "Pessach I"
  ["2021-03-28"]=>
  string(9) "Pessach I"
  ["2022-04-16"]=>
  string(9) "Pessach I"
}
*/

/*
 * get Config from a URL
 */
$url = "http://example.com/data/JspitHoliday.sqlite";
$tmpfname = tempnam(sys_get_temp_dir(), "holiday.sqlite");
$copyOk = copy($url,$tmpfname);

$holidaysDE = new JspitHoliday('de',$tmpfname);

```
## Class-Info

| Info | Value |
| :--- | :---- |
| Declaration | class JspitHoliday |
| Datei | JspitHoliday.php |
| Date/Time modify File | 2018-08-09 15:23:52 |
| File-Size | 20 KByte |
| MD5 File | 5796df9354e50ad44ef21df9930211d0 |
| Version | 1.30 |
| Date | 2018-08-09 |

## Public Methods

| Methods and Parameter | Description/Comments |
| :-------------------- | :------------------- |
| public function __construct($filterRegion = &quot;&quot;, $db = null, $typFilter = self::TYPE_ALL) | Constructs the class instance<br>@param string $filterRegion Country/Region ISO 3361 Alpha2 (&#039;DE&#039;,&#039;DE-BY&#039;..) <br>@param string $db filename for SQLite or PDO Object, default: holiday.sqlite<br>@param int $typFilte Filter for Holiday-Type for SQLite, default: holiday::TYPE_ALL<br>@throws InvalidArgumentException |
| public static function create($filterRegion = &quot;&quot;, $sqliteFile = null, $typFilter = self::TYPE_ALL) | return a new class instance<br>@param string $filterRegion Country/Region ISO 3361 Alpha2 (&#039;DE&#039;,&#039;DE-BY&#039;..) <br>@param string $sqliteFile filename for SQLite, default: holiday.sqlite<br>@param int $typFilte Filter for Holiday-Type for SQLite, default: holiday::TYPE_ALL<br>@return object JspitHoliday<br>@throws InvalidArgumentException |
| public function setLanguage($language = &quot;en-GB&quot;) | set Default Language<br>@param string $language p.E. &quot;de-DE&quot;, &quot;en-GB&quot; <br>@return $this |
| public function getLanguage() | get Default Language<br>@return string default language |
| public function setRegion($filterRegion) | set region<br>@param string $filterRegion Country/Region ISO 3361 Alpha2 (&#039;DE&#039;,&#039;DE-BY&#039;..) <br>@return $this |
| public function getRegion() | Returns the current Region<br>@return string |
| public function setTypFilter($typFilter = self::TYPE_ALL) | set Filter Holiday Type<br>@param int typ Filter<br>@throws Exception |
| public function holidayName($date = &quot;today&quot;, $language = null) | get Name from a Holiday p.e: &quot;New Year&#039;s Day&quot;<br>@param $date: string, datetime-object or timestamp <br>@param $language string p.E. &quot;de-DE&quot;, &quot;en-GB&quot; <br>@return mixed string name if ok, false Error or Date is not a Holiday,<br>string &quot;?&quot; no Name for the language in Database  |
| public function holidayList($year = null, $language = null) | return array( &#039;YYYY-MM-DD&#039; =&gt; holidayname, ..)<br>the array is sorted by ascending date<br>@param integer year full year p.E. 2018<br>@param string $language p.E. &quot;en_GB&quot;<br>@return array |
| public function dateList($startDate, $endDate, $weekDayFilterList = null) | returns array of string dates Y-m-d <br>wich are a holiday between two dates<br>@param mixed $startDate<br>@param mixed $endDate<br>@param array $weekDayFilterList with numbers 0..6 for Sunday ..Saturday<br>default null for all weekdays <br>@return array of strings [&quot;Y-m-d&quot;,..] |
| public function dateTimeList($year = null, $language = null) | return array of datetime objects<br>the array is sorted by ascending date<br>datetime objects are extended with public property holidayName<br>@param year integer full year p.E. 2018<br>@param $language string p.E. &quot;en_GB&quot;<br>@return array |
| public function isHoliday($date = &#039;today&#039;) | return true id if date is a holiday or false<br>@param mixed $date string, datetime-object or timestamp <br>@return bool |
| public function getNames($nameFilter = &quot;&quot;, $language = null, $onlyCurrentRegion = false) | get List of Names from DB as <br>array(idholiday =&gt; name, ..) by nameFilter and language<br>return false if not found<br>@param string $nameFilter Filter for name , caseinsenitive <br>@param string $language how de or de-ch, default Default Language<br>@param bool $onlyCurrentRegion bool, default false<br>@return mixed  |
| public function holidayNameList($nameFilter = &quot;&quot;,$yearStart = null, $countYears=1, $language = null) | get List of Names from DB as <br>array(idholiday =&gt; name, ..) by nameFilter and language<br>return false if not found<br>@param nameFilter: Filter for name , caseinsenitive <br>@param yearStart: integer YYYY , default current year<br>@param countYears: default 1 or end of year (end of year &gt; <br>@param language string how de or de-ch, default Default Language<br>@return mixed |
| public function getConfig() | get config, may use as debugging info<br>@return array |
| public static function getEasterDate($year,$orthodox = false) | get easter-date as string YYYY-MM-DD |
| public static function getPassoverDate($year) | calculate the first day of Passover (Gauß)<br>@params: $year integer as YYYY, interval 1900 to 2099<br>@return date as string YYY-MM-DD |
| public static function getHijriDate($gregYear, $hijriMonth, $hijriDay) | get the gregorian Date for the year $gregYear<br>@param integer $gregYear: greg.Year (2007..2031)<br>@param integer $hijriMonth: Month hijri-Calendar<br>@param integer $hijriMonth: Month hijri-Calendar<br>@return date as string YYY-MM-DD or false if error |
| public static function GregToHijri($y,$m,$d) | convert Gregorian Date to Hijri<br>return array($year, $month, $day) |
| public static function HijriToGreg($y,$m,$d) | convert Hijri Date to Gregorian<br>return array($year, $month, $day) |
| public static function getJewishDate($gregYear, $jewishMonth, $jewishDay) | return string date YYYY-MM-DD; |
| public static function getCalendarDate($calendar, $calMonth, $calDay, $gregYear) | get greg.Date from $calMonth and $calDay in calendar $cal<br>return string date &quot;Y-m-d&quot; or false if error |

## Constants

| Declaration/Name | Value | Description/Comments |
| :--------------- | :---- | :------------------- |
|  const TYPE_OFFICIAL = 1; | 1 |   |
|  const TYPE_BANK = 2; | 2 |   |
|  const TYPE_OBSERVED = 4; | 4 |   |
|  const TYPE_OTHER = 8; | 8 |   |
|  const TYPE_4 = 16; | 16 |   |
|  const TYPE_5 = 32; | 32 |   |
|  const TYPE_6 = 64; | 64 |   |
|  const TYPE_ALL = 0x7FFF; | 32767 |   |

## Define country depending holidays

All holidays are dates defined in the table 'holidays' and names for all languages in the table 'names'.
Working with SQLite database will be easy if you use a tool like DB Browser for SQLite ( http://sqlitebrowser.org ) and
the test enviroment phpcheck. Download as zip and unzip in a public directory.
Then you can call phpcheck.JspitHolidayOffice.php in the browser. Edit the Source of phpcheck.JspitHolidayOffice.php. 
Take the big list of countries in comment and use your country for that. 
Check after each new database entry if the holiday date is correctly determined.

### Fields of  holidays table:

| Field | Description |
| ----- | ----------- |
| id | id, autoincrement, reference to 'idholiday' in the table names |
| comment | a comment (not a name for a holiday) |
| year | free or "*" for all years, a year YYYY for only this year, -YYYY to year, YYYY- from year, a range YYYY-YYYY, a list of years YYYY,YYYY,.. |
| except_year | free for no exception, YYYY for except only this year, except a range YYYY-YYYY, except a list of years YYYY,YYYY,.., "*" except all |
| month | used for fixed months |
| day | used for fixed days |
| special | A pipe with relative date formates and wildcards. Pipe elements are are separated by \| . {{name}} is a wildcard. Some examples: "first sunday of september {{year}}\|next thursday" ,'third sunday of september {{year}}' , '{{easter}}\|+1 Day' |
| region | A list auf Countrycodes/Regions. Countrycode-[[[Subdivision]-Subregion1]-Subregion2] |
| typ | Type of holiday (TYPE_OFFICIAL, TYPE_BANK..) |

### Fixed date every year

Table holidays

| id  | comment     | year | except_year | month | day | special | region      | typ |
| --- | ----------- | ---- | ----------- | ----- | --- | ------- | ----------- | --- | 
| 1   | NewYear     |      |             | 1     | 1   |         | DE,CH,AT,NL | 1   |

Table names

changelog.md
# JspitHoliday
## Version 1.32 (2023-04-18)
* tested for php 8.2

## Version 1.31 (2019-10-21)
* add method dateList

## Version 1.30 (2018-08-03)
* calendars implemented 



class/JspitHoliday.php
<?php
/**
.---------------------------------------------------------------------------.
|  Software: JspitHoliday - PHP class                                       |
|   Version: 1.32                                                           |
|      Date: 2023-04-18                                                     |
| ------------------------------------------------------------------------- |
| Copyright © 2018,2019 Peter Junk alias jspit All Rights Reserved.         |
'---------------------------------------------------------------------------'
*/
#[AllowDynamicProperties] 

class JspitHoliday
{
  const TYPE_OFFICIAL = 1;
  const TYPE_BANK = 2;
  const TYPE_OBSERVED = 4;
  const TYPE_OTHER = 8;
  const TYPE_4 = 16;
  const TYPE_5 = 32;
  const TYPE_6 = 64;
  const TYPE_ALL = 0x7FFF;
  
  protected $pdo;
  protected $language = "en-GB";
  protected $region;
  protected $config;
  protected $typFilter;

  /**
   * Constructs the class instance
   * @param string $filterRegion Country/Region ISO 3361 Alpha2 ('DE','DE-BY'..) 
   * @param string $db filename for SQLite or PDO Object, default: holiday.sqlite
   * @param int $typFilte Filter for Holiday-Type for SQLite, default: holiday::TYPE_ALL
   * @throws InvalidArgumentException
   */
  public function __construct($filterRegion = "", $db = null, $typFilter = self::TYPE_ALL) {
    //verify filter
    $filter = strtoupper($filterRegion);
    if(!preg_match('/^[A-Z]{2,3}(-[A-Z0-9]{1,8}){0,3}\-?\*?$/', $filter)) {
      throw new InvalidArgumentException("filterRegion is not like ISO3361");  
    }
     
    if($db instanceof PDO) {
      try{  
        $this->pdo = $db;
        $this->createConfig($filterRegion, $typFilter);
      } catch(Exception $e) {
        throw new InvalidArgumentException("Faulty PDO-Connection");  
      }
    }
    else {    
      if(! is_string($db) OR $db == "") {
        $db = __DIR__ . "/".basename(__CLASS__).".sqlite";
      }
      if(!file_exists($db)){
        throw new InvalidArgumentException("SQLite File '$db' not found");
      }
      try{
        $options = array(
          PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
          PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_OBJ,
        );

        $this->pdo = new PDO('sqlite:'.$db,null,null,$options);
        
        $this->createConfig($filterRegion, $typFilter);
      } catch(Exception $e) {
        throw new InvalidArgumentException("Faulty SQLite-DB '$db'");  
      }
    }
  }
  
 /**
  * return a new class instance
  * @param string $filterRegion Country/Region ISO 3361 Alpha2 ('DE','DE-BY'..) 
  * @param string $sqliteFile filename for SQLite, default: holiday.sqlite
  * @param int $typFilte Filter for Holiday-Type for SQLite, default: holiday::TYPE_ALL
  * @return object JspitHoliday
  * @throws InvalidArgumentException
  */
  public static function create($filterRegion = "", $sqliteFile = null, $typFilter = self::TYPE_ALL) {
    return new static($filterRegion, $sqliteFile, $typFilter );    
  }
  
  
 /**
  * set Default Language
  * @param string $language p.E. "de-DE", "en-GB" 
  * @return $this
  */
  public function setLanguage($language = "en-GB") {
    $this->language = $language;
    return $this;
  }

 /**
  * get Default Language
  * @return string default language
  */
  public function getLanguage() {
    return $this->language;
  }

 /**
  * set region
  * @param string $filterRegion Country/Region ISO 3361 Alpha2 ('DE','DE-BY'..) 
  * @return $this
  */
  public function setRegion($filterRegion) {
    try{
      $this->createConfig($filterRegion, $this->typFilter);
      return $this;
    } catch(Exception $e) {
      throw $e;
    }
  }
  
 /**
  * Returns the current Region
  * @return string
  */
  public function getRegion() {
    return $this->region;
  }
  
 /**
  * set Filter Holiday Type
  * @param int typ Filter
  * @throws Exception
  */
  public function setTypFilter($typFilter = self::TYPE_ALL) {
    try{
      $this->createConfig($this->region, (int)$typFilter);
      return $this;
    } catch(Exception $e) {
      throw $e;
    }
  }

 /**
  * get Name from a Holiday p.e: "New Year's Day"
  * @param $date: string, datetime-object or timestamp 
  * @param $language string p.E. "de-DE", "en-GB" 
  * @return mixed string name if ok, false Error or Date is not a Holiday,
  *  string "?" no Name for the language in Database  
  */
  public function holidayName($date = "today", $language = null){
    if(is_string($date)) {
      $date = date_create($date)->format("Y-m-d");
    } elseif($date instanceof DateTime) {
      $date = $date->format("Y-m-d");
    } elseif(is_int($date)) {
      $date = date("Y-m-d", $date);
    } else {
      throw new InvalidArgumentException("incorrect Parameter date '$date' "); 
    }
    
    $id = $this->getId($date);
    if(!$id) return false;  //holiday by date not found
    
    if($language === null) $language = $this->language;
    $name = $this->getHolidayNameById($id, $language);

    return $name;
  }

 /**
  * return array( 'YYYY-MM-DD' => holidayname, ..)
  * the array is sorted by ascending date
  * @param integer year full year p.E. 2018
  * @param string $language p.E. "en_GB"
  * @return array
  */
  public function holidayList($year = null, $language = null){
    if(empty($year)) $year = date('Y');
    if($year < 1900 OR $year > 2099){
      throw new InvalidArgumentException("incorrect Parameter year '$year' "); 
    }
    if($language === null) $language = $this->language;

    $hList = array();
    foreach($this->config as $id => $row){
      $curDate = $this->getDateFromDBrow($row, $year);
      if($curDate === false) continue;

      $hList[$curDate] = $this->getHolidayNameById($id, $language);
    }
    ksort($hList); 
    return $hList;
  }

  /**
   * returns array of string dates Y-m-d 
   * wich are a holiday between two dates
   * @param mixed $startDate
   * @param mixed $endDate
   * @param array $weekDayFilterList with numbers 0..6 for Sunday ..Saturday
   *   default null for all weekdays 
   * @return array of strings ["Y-m-d",..]
   */
  public function dateList($startDate, $endDate, $weekDayFilterList = null)
  {
    if(is_string($startDate)) {
      $startDate = date_create($startDate)->format("Y-m-d");
    } elseif($startDate instanceof DateTime) {
      $startDate = $startDate->format("Y-m-d");
    } elseif(is_int($startDate)) {
      $startDate = date("Y-m-d", $startDate);
    } else {
      throw new InvalidArgumentException("incorrect Parameter date '$startDate' "); 
    }

    if(is_string($endDate)) {
      $endDate = date_create($endDate)->format("Y-m-d");
    } elseif($endDate instanceof DateTime) {
      $endDate = $endDate->format("Y-m-d");
    } elseif(is_int($endDate)) {
      $endDate = date("Y-m-d", $endDate);
    } else {
      throw new InvalidArgumentException("incorrect Parameter date '$endDate' "); 
    }

    $startYear = (int)substr($startDate,0,4);
    $endYear = (int)substr($endDate,0,4);
    $dateArr = array();
    foreach($this->config as $id => $row){
      for($year = $startYear; $year <= $endYear; $year++){
        $curDate = $this->getDateFromDBrow($row, $year);
        if($curDate !== false AND $curDate >= $startDate AND $curDate <= $endDate){
          if(is_array($weekDayFilterList)){
            $numberWeekDay = (int)date_create($curDate)->format('w');
            if(!in_array($numberWeekDay,$weekDayFilterList)) continue;
          }
          $dateArr[] = $curDate;
        } 
      }
    }
    sort($dateArr);
    return $dateArr;
  }
  
 /**
  * return array of datetime objects
  * the array is sorted by ascending date
  * datetime objects are extended with public property holidayName
  * @param year integer full year p.E. 2018
  * @param $language string p.E. "en_GB"
  * @return array
  */
  public function dateTimeList($year = null, $language = null){
    $dtArr = array();
    foreach(self::holidayList($year, $language) as $strDate => $name) {
      $dt = date_create($strDate);
      if(is_object($dt)) {
        $dt->holidayName = $name;
        $dtArr[] = $dt;
      }
    }
    return $dtArr;
  }


 /**
  * return true id if date is a holiday or false
  * @param mixed $date string, datetime-object or timestamp 
  * @return bool
  */
  public function isHoliday($date = 'today'){
    if(is_string($date)) {
      $date = date_create($date)->format("Y-m-d");
    } elseif($date instanceof DateTime) {
      $date = $date->format("Y-m-d");
    } elseif(is_int($date)) {
      $date = date("Y-m-d", $date);
    } else {
      throw new InvalidArgumentException("date '$date' incorrect"); 
    }
    
    return $this->getId($date) ? true : false;
    
  }
  
 /**
  * get List of Names from DB as 
  * array(idholiday => name, ..) by nameFilter and language
  * return false if not found
  * @param string $nameFilter Filter for name , caseinsenitive 
  * @param string $language  how de or de-ch, default Default Language
  * @param bool $onlyCurrentRegion bool, default false
  * @return mixed  
  */
  public function getNames($nameFilter = "", $language = null, $onlyCurrentRegion = false) {
    if($language === null) $language = $this->language;
    $sql = "SELECT idholiday, name
      FROM names 
      WHERE language LIKE :language COLLATE NOCASE"; 
      $param = array("language" => "%".$language."%");
    if($nameFilter != "") {
      $sql .= " AND name LIKE :nameFilter COLLATE NOCASE";
      $param['nameFilter'] = "%".$nameFilter."%";
    }
    $stmt = $this->pdo->prepare($sql);
    $stmt->execute($param);
    $rows = $stmt->fetchAll(PDO::FETCH_KEY_PAIR);
    if($rows AND $onlyCurrentRegion) {
      $rows = array_intersect_key($rows, $this->config);  
    }
    return $rows ? $rows : false; 
  }
  
 /**
  * get List of Names from DB as 
  * array(idholiday => name, ..) by nameFilter and language
  * return false if not found
  * @param nameFilter: Filter for name , caseinsenitive 
  * @param yearStart: integer YYYY , default current year
  * @param countYears:  default 1 or end of year (end of year > 
  * @param language string how de or de-ch, default Default Language
  * @return mixed
  */
  public function holidayNameList($nameFilter = "",$yearStart = null, $countYears=1, $language = null) {
    if($yearStart === null) $yearStart = date("Y");
    //countYears or end of year
    if($countYears > 1000 AND $countYears >= $yearStart){
       $countYears = $countYears - $yearStart + 1; 
    }
    //get id => name array
    $idNames = $this->getNames($nameFilter, $language, true);
    if($idNames === false) return false;

    $list = array();
    for($i=0; $i < $countYears; $i++){
      $year = $yearStart + $i;
      foreach($idNames as $id => $HolidayName) {
        if(array_key_exists($id, $this->config)) {
          $curDate = $this->getDateFromDBrow($this->config[$id], $year);
          if($curDate) $list[$curDate] = $HolidayName;
        } else {  
        //id not in current region
        }
      }
    }  
    ksort($list);
    return $list;
  }
 
 /** 
  * get config, may use as debugging info
  * @return array
  */
  public function getConfig(){
    return $this->config;
  }
  
  // get id from holiday-Table
  // @param $strDate string format "YYYY-MM-DD"
  protected function getId($date){
    list($year,$month,$day) = explode("-",$date);
    foreach($this->config as $id => $row){
      if($date == $this->getDateFromDBrow($row, $year, $month, $day)) {
        return $id;
      }
    }
    return false;
  }
  
 /*
  * get date as string YYYY-MM-DD
  * $code string : special Code 
  * $year 1600 < year < 2100
  * return false if error
  */
  protected function getMovableDate($code, $year, $month = 1, $day = 1){
    if(empty($code)) {
      return sprintf("%04d-%02d-%02d",$year, $month, $day);
    }
    if(strpos($code,'}}')) {
      $callback = function($match) use($year, $month, $day){
        $m0 = $match[0];
        $c = __CLASS__;
        if($m0 == '{{year}}') return $year;
        if($m0 == '{{month}}') return $month;
        if($m0 == '{{day}}') return $day;
        if($m0 == '{{easter}}') return $c::getEasterDate($year);
        if($m0 == '{{easter_o}}') return $c::getEasterDate($year, true);
        if($m0 == '{{passover}}') return $c::getPassoverDate($year);
        if($m0 == '{{islamic}}') return $c::getHijriDate($year, $month, $day);
        if($m0 == '{{hebrew}}') return $c::getJewishDate($year, $month, $day);  //jewish
        if(extension_loaded('intl')) {
          $intlCalendars = array(
            '{{japanese}}','{{buddhist}}','{{chinese}}','{{persian}}',
            '{{indian}}','{{coptic}}','{{ethiopic}}'
          );
          if(in_array($m0,$intlCalendars)) {
            return $c::getCalendarDate(trim($m0,'{}'), $month, $day, $year);  
          }
        }
        return $m0;  //not replace
      };
      $code = preg_replace_callback('~\{\{[a-z_]+\}\}~', $callback, $code);
      
      
      //check extends methods
      if(preg_match('~\{\{([a-z]+)\}\}~',$code,$match)) {
        $methodName = $match[1];
        if(method_exists($this, $methodName)) {
          $replacement = $this->$methodName($year, $month, $day);
          $code = str_replace($match[0],$replacement,$code);
        } else { //error
          throw new Exception("Error ".__CLASS__.": unknown special entry ".$match[0]); 
        }      
      }
    }
    
    $modifiers = explode("|", $code);
    $date = date_create($year."-".$month."-".$day);
    foreach($modifiers as $modify) {
      //check for datelist {{2018:2/3,..}}
      if(preg_match('~\{\{(\d{4}):(.*)\}\}~',$modify,$match)) {
        $startYear = $match[1];
        $values = explode(",",$match[2]);
        $key = $year - $startYear; 
        if(array_key_exists($key,$values)) {
          $modify = str_replace($match[0],$values[$key],$modify);
        } else {
          //if year not exist in list get static day and month
          $modify = $month."/".$day;
        }
        $date->modify($modify);
      }

      elseif(preg_match('~^\{\{\?([DdmLY]+)(!?=)([^}]+)\}\}(.*)~',$modify,$match)) {
        $curFmt = $date->format($match[1]);
        $found = stripos($match[3], $curFmt) !== false;
        if($found === ($match[2] == "=")) {
          //condition true
          if($match[4] !== "") $date->modify($match[4]);
        } elseif($match[4] === "") {
          return false;
        }
      } else { 
        $date->modify($modify);
      }        
      $errArr = date_get_last_errors();
      if($errArr && $errArr['error_count']) return false;
    }
    
    return $date->format("Y-m-d");
   }
  
  //get easter-date as string YYYY-MM-DD
  public static function getEasterDate($year,$orthodox = false){
    if($orthodox) {
      $flag = CAL_EASTER_ALWAYS_JULIAN;
      $basisDate = $year."-4-3";
    } else {
      $flag = CAL_EASTER_ALWAYS_GREGORIAN;
      $basisDate = $year."-3-21";
    }
    $date = date_create($basisDate)
      ->modify(easter_days($year, $flag).' Days')
    ;
    return $date
      ->modify((-(int)$date->format('w'))." Days")
      ->format("Y-m-d");
  }
  
 /*
  * calculate the first day of Passover (Gauß)
  * @params: $year integer as YYYY, interval 1900 to 2099
  * @return date as string YYY-MM-DD
  */
  public static function getPassoverDate($year){
    $a = (12*$year+12)%19; 
    $b = $year%4;
    $m = 20.0955877 + 1.5542418 * $a + 0.25 * $b - 0.003177794 * $year; 
    $mi = (int)$m;
    $mn = $m-$mi;
    $c = ($mi + 3 * $year + 5 * $b + 1)%7; 
    if($c==2 OR $c==4 OR $c==6) {
      $mi += 1;
    } elseif($c==1 AND $a > 6 AND $mn >= (1367/2160)) {
      $mi += 2;
    } elseif ($c==0 AND $a > 11 AND $mn > (23269/25920)) {
      $mi += 1;
    }
    return date_create($year."-3-13")
      ->modify($mi." Days")
      ->format("Y-m-d"); 
  }
  
 /* 
  * get the gregorian Date for the year $gregYear
  * @param integer $gregYear: greg.Year (2007..2031)
  * @param integer $hijriMonth: Month hijri-Calendar
  * @param integer $hijriMonth: Month hijri-Calendar
  * @return date as string YYY-MM-DD or false if error
  */
  public static function getHijriDate($gregYear, $hijriMonth, $hijriDay) {
    $gy = $gregYear;
    for($i=0;$i<3;$i++){
      $hijri = self::GregToHijri($gy,1,1);
      $greg  = self::HijriToGreg($hijri[0],$hijriMonth,$hijriDay);
      if($greg[0] == $gregYear) {
        return sprintf("%04d-%02d-%02d",$greg[0],$greg[1],$greg[2]);
      }
      $gy += ($greg[0] < $gy) ? 1 : -1;
    }
    return false;
  }  

 /*
  * convert Gregorian Date to Hijri
  * return array($year, $month, $day)
  */  
  public static function GregToHijri($y,$m,$d)
  {
      $jd = gregoriantojd ($m, $d, $y);
      $days360month = 10631;
      $y = $days360month / 30.0;
      $shift1 = 8.01 / 60.0;
      $z = $jd - 1948085; //-$epochAstro
      $cyc = floor($z / $days360month);
      $z = $z - $days360month * $cyc;
      $j = floor(($z - $shift1) / $y);
      $z = $z - floor($j * $y + $shift1);
      $year = 30 * $cyc + $j;
      $month = (int)floor(($z + 28.5001) / 29.5);
      if ($month == 13) $month = 12;
      $day = $z - floor(29.5001 * $month - 29);
      return array($year,$month,$day);
  }

 /*
  * convert Hijri Date to Gregorian
  * return array($year, $month, $day)
  */  
  public static function HijriToGreg($y,$m,$d){
   $jd = (int)((11 * $y + 3) / 30) + 354 * $y + 
     30 * $m - (int)(($m - 1) / 2) + $d + 1948440 - 385;
   list($month,$day,$year) = explode("/", jdtogregorian($jd));
   return array($year,$month,$day);
  }
  
 /*
  * return string date YYYY-MM-DD;
  */
  public static function getJewishDate($gregYear, $jewishMonth, $jewishDay) {
    $gy = $gregYear;
    for($i=0;$i<3;$i++){
      $jd1 = gregoriantojd(1,1,$gy);
      $jw1 = jdtojewish($jd1);  //"Monat/Tag/Jahr"
      list($jmonth,$jday,$jyear) = explode("/",$jw1);
      $jd = jewishtojd($jewishMonth,$jewishDay,$jyear);
      $greg = jdtogregorian($jd);
      list($gmonth,$gday,$gyear) = explode("/",$greg);
      if($gyear == $gregYear) {
        return sprintf("%04d-%02d-%02d",$gyear,$gmonth,$gday);
      }
      $gy += ($gyear < $gy) ? 1 : -1;
    }
    return false;
  }

 /*
  * get greg.Date from $calMonth and $calDay in calendar $cal
  * return string date "Y-m-d" or false if error
  */
  public static function getCalendarDate($calendar, $calMonth, $calDay, $gregYear){
    if(!extension_loaded('intl')) return false;
    $formatter = IntlDateFormatter::create(
      'de_DE', 
      IntlDateFormatter::FULL, 
      IntlDateFormatter::FULL,
      null, 
      IntlDateFormatter::GREGORIAN,
      'yyyy-MM-dd'
    );
    
    $cal = IntlCalendar::createInstance(null, "@calendar={$calendar}");
    
    $gy = $gregYear;
    for($i=0;$i<3;$i++){
      $msTs = mktime(0,0,0,1,1,$gy) * 1000.0;
      $cal->setTime($msTs);
      $cal->set(IntlCalendar::FIELD_MONTH, $calMonth-1); //month-1 !
      $cal->set(IntlCalendar::FIELD_DAY_OF_MONTH, $calDay); 
      $gregDate = $formatter->format($cal);
      $gyear = (int)substr($gregDate,0,4);
      if($gyear == $gregYear) return $gregDate;
      $gy += ($gyear < $gy) ? 1 : -1;
    }
    return false;  
  }  
  



 /*
  * return data as string YYYY-MM-DD or false
  */
  protected function getDateFromDBrow($row, $year, $month = 1, $day = 1){
    //accept years
    if(strlen($row->year) >= 4) {
      $dbEntry = $row->year;
      if(ctype_digit($dbEntry)) {
        //only YYYY
        if($dbEntry != $year) return false;
      } elseif(preg_match('~^\d{4}-$~',$dbEntry)) {
        //YYYY-
        if($year < (int)$dbEntry) return false;
      } elseif(preg_match('~^-\d{4}$~',$dbEntry)) {
        //-YYYY
        if($year > (-(int)$dbEntry)) return false;
      } elseif(!in_array($year,$this->listToArray($dbEntry))) {
        return false;
      }
    }
        
    if(strlen($row->except_year) >= 1) {
      if($row->except_year == '*') return false;
      if(in_array($year,$this->listToArray($row->except_year))){
        return false;
      }
    }

    $curmonth = $row->month ? $row->month : $month;
    $curday = $row->day ? $row->day : $day;
    if($row->special) {
      //getMovableDate return false if not match or error
      $curDate = $this->getMovableDate($row->special,$year,$curmonth,$curday);
    } else {
      $curDate = sprintf("%04d-%02d-%02d",$year,$curmonth,$curday);
    }
    return $curDate;
  }
  
  //create config-array
  private function createConfig($region, $typFilter){
    $filterRegion = preg_replace('~\-?\*$~','',$region);
    $allOption = $filterRegion != $region; //de*
    $sql = "SELECT id, year, except_year, month, day, special, region
      FROM holidays 
      WHERE typ & ".(int)$typFilter.
      " ORDER BY except_year DESC, year DESC";

    $stmt = $this->pdo->query($sql);
    
    $this->config = array();
    $match = false;
    foreach($stmt as $row){
      $dbRegios = explode(",",$row->region);
      foreach($dbRegios as $region) {
        if($allOption) {
          if ($match = (stripos($region,$filterRegion) === 0)) break;
        } else {       
          if($match = (stripos($filterRegion,$region) === 0)) break;
        }
      }
      if($match) $this->config[$row->id] = $row;
    }
    $this->region = $filterRegion;
    $this->typFilter = $typFilter;
    return true;
  }

  //string list to array
  private function listToArray($strList){
    $strList = preg_replace_callback(
      '/(\d{4})-(\d{4})/',
      function(array $m){
        return implode(",",range($m[1],$m[2]));
      },
      $strList
    );
    return explode(",",$strList);  
  }

  //get Name by id and $language
  private function getHolidayNameById($id, $language){
    $sql = "SELECT name 
      FROM names 
      WHERE idholiday = $id AND language LIKE :language COLLATE NOCASE
      LIMIT 1";
    $stmt = $this->pdo->prepare($sql);
    $stmt->execute(array("language" => "{$language}%"));
    $row = $stmt->fetch();
    if($row) return $row->name;
    return "?";
  }
  
  
  
}  
